lib/solarwinds_apm/support/resource_detector.rb

```diff
@@ -0,0 +1,129 @@
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+require 'net/http'
+require 'uri'
+require 'json'
+
+module SolarWindsAPM
+  # ResourceDetector
+  # Usage:
+  #   require 'opentelemetry/sdk'
+  #   require 'opentelemetry/resource/detector'
+  #   OpenTelemetry::SDK.configure do |c|
+  #     c.resource = SolarWindsAPM::ResourceDetector.detect
+  #   end
+  module ResourceDetector
+    K8S_PODNAME_PATH = '/etc/hostname'
+    K8S_NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+    K8S_TOKEN_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/token'
+
+    UAMS_CLIENT_PATH = '/opt/solarwinds/uamsclient/var/uamsclientid'
+    UAMS_CLIENT_URL = 'http://127.0.0.1:2113/info/uamsclient'
+    UAMS_CLIENT_ID_FIELD = 'uamsclient_id'
+
+    def self.detect
+      attributes = ::OpenTelemetry::SDK::Resources::Resource.create({})
+      attributes = attributes.merge(detect_uams_client_id)
+      attributes = attributes.merge(detect_k8s_atttributes)
+      attributes.merge(from_upstream_detector)
+    end
+
+    def self.detect_uams_client_id
+      uams_client_id = nil
+      if File.exist?(UAMS_CLIENT_PATH)
+        uams_client_id = File.read(UAMS_CLIENT_PATH).strip
+      else
+        url = URI(UAMS_CLIENT_URL)
+
+        response = nil
+        ::OpenTelemetry::Common::Utilities.untraced do
+          http = Net::HTTP.new(url.host, url.port)
+          request = Net::HTTP::Get.new(url)
+          response = http.request(request)
+        end
+
+        if response&.code.to_i == 200
+          uams_metadata = JSON.parse(response.body)
+          uams_client_id = uams_metadata&.fetch(UAMS_CLIENT_ID_FIELD)
+        end
+      end
+
+      resource_attributes = {
+        'sw.uams.client.id' => uams_client_id
+      }
+
+      ::OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+    rescue StandardError => e
+      SolarWindsAPM.logger.debug "#{self.class}/#{__method__}] detect_uams_client_id failed. Error: #{e.message}."
+      ::OpenTelemetry::SDK::Resources::Resource.create({})
+    end
+
+    def self.detect_k8s_atttributes
+      return ::OpenTelemetry::SDK::Resources::Resource.create({}) unless ENV['KUBERNETES_SERVICE_HOST'] && ENV['KUBERNETES_SERVICE_PORT']
+
+      resource_attributes = {}
+      pod_name = File.read(K8S_PODNAME_PATH).strip if File.exist?(K8S_PODNAME_PATH)
+      namespace = File.read(K8S_NAMESPACE_PATH).strip if File.exist?(K8S_NAMESPACE_PATH)
+      pod_uid = nil
+      token = nil
+
+      token = File.read(K8S_TOKEN_PATH).strip if File.exist?(K8S_TOKEN_PATH)
+
+      if token && pod_name && namespace
+        url = URI("https://kubernetes.default.svc/api/v1/namespaces/#{namespace}/pods/#{pod_name}")
```

Comment:
We cannot use the Kubernetes API to retrieve k8s metadata, because customers likely do not want to and/or cannot set up the token and required permissions. Take a look at our [spec](https://github.com/solarwinds-cloud/trace/blob/resource-attributes/docs/Collector-RPC-API.md#k8smetadata) and you can reference the JS implementation https://github.com/solarwinds/apm-js/blob/main/packages/instrumentations/src/resource-detector-k8s.ts.

Reply:
Updated to not use private api get metadata